### PR TITLE
add artifact URL as soon as reconciling starts

### LINKED ISSUE #144: Provide GCS URL once model is created instead of once ready
Right now the GCS url is only provided once the model is ready, however it would be nice to be able to pull down the logs as the model is being loaded/trained. So hoping that if we provide the URL I can pull down logs directly while it's loading.

### PR DESCRIPTION
**Pull request by samos123**
Fixes #144

### CODE REVIEW

File: internal/controller/model_controller.go [hunk 1]
```diff
@@ -68,10 +68,23 @@ func (r *ModelReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 func (r *ModelReconciler) reconcileModel(ctx context.Context, model *apiv1.Model) (result, error) {
 	log := log.FromContext(ctx)
 
-	if model.Status.Artifacts.URL != "" {
+	if model.Status.Ready {
```
Comment (nstogner) [inline internal/controller/model_controller.go:71]: Keep this change to the conditional

File: internal/controller/model_controller.go [hunk 2]
```diff
@@ -68,10 +68,23 @@ func (r *ModelReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
...
 		return result{success: true}, nil
 	}
 
+	if model.Status.Artifacts.URL == "" {
+		model.Status.Artifacts.URL = r.Cloud.ObjectArtifactURL(model).String()
```
Comment (nstogner) [inline internal/controller/model_controller.go:76]: Always set this URL.

File: internal/controller/model_controller.go [hunk 3]
```diff
@@ -68,10 +68,23 @@ func (r *ModelReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
...
+		meta.SetStatusCondition(model.GetConditions(), metav1.Condition{
+			Type:               apiv1.ConditionReconciling,
+			Status:             metav1.ConditionTrue,
+			Reason:             apiv1.ReasonJobComplete,
+			ObservedGeneration: model.Generation,
+		})
+		if err := r.Status().Update(ctx, model); err != nil {
```
Comment (nstogner) [inline internal/controller/model_controller.go:83]: Remove this status update... This will be too fleeting, not worth the API request.

Comment (nstogner) [reply internal/controller/model_controller.go:83]: This update request will be sent very soon after this and the URL will be set then.

File: api/v1/conditions.go [hunk 4]
```diff
@@ -1,11 +1,12 @@
 package v1
 
 const (
-	ConditionUploaded = "Uploaded"
-	ConditionBuilt    = "Built"
-	ConditionLoaded   = "Loaded"
-	ConditionModelled = "Modelled"
-	ConditionDeployed = "Deployed"
+	ConditionReconciling = "Reconciling"
```
Comment (nstogner) [inline api/v1/conditions.go:4]: Conditions here should be end-states not transition states. IMO, I would remove this and go with the suggested changes below which make this unnecessary.

**Review by nstogner — CHANGES_REQUESTED** [review]
Agree we should set URL earlier, but I don't think we want to include a new call to update status. I would just set the field and allow the next update status call to do what you are intending here.

File: internal/controller/model_controller.go (hunk 1 above)
Comment (samos123) [reply internal/controller/model_controller.go:71]: you mean change model.Status.Ready == true?

Comment (samos123) [reply internal/controller/model_controller.go:71]: I think what I have currently is preferred over having
```
if model.Status.Ready == true {
```

Source: https://stackoverflow.com/questions/3685002/check-if-boolean-is-true

**Comment by samos123** [comment]
I verified the new approach of not using an additional update works perfectly as well. Thanks for the great suggestion

File: internal/controller/model_controller.go (hunk 1 above)
Comment (nstogner) [reply internal/controller/model_controller.go:71]: Was just commented on checking ready instead of the url

**Review by nstogner — APPROVED** [review]
Looks good!

**Review by brandonjbjelland — APPROVED** [review]

File: internal/controller/model_controller.go (hunk 1 above)
Comment (samos123) [reply internal/controller/model_controller.go:71]: not sure why I misunderstood your comment. I read it as change to the conditional format or smth like that.

**Pull request merged**